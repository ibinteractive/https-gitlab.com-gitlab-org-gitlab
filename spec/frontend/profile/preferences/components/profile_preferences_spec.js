import { GlSprintf } from '@gitlab/ui';
import { shallowMount } from '@vue/test-utils';

import ProfilePreferences from '~/profile/preferences/components/profile_preferences.vue';
import IntegrationView from '~/profile/preferences/components/integration_view.vue';
import {
  languageChoices,
  firstDayOfWeekChoicesWithDefault,
  integrationViews,
  userFields,
  featureFlags,
} from '../mock_data';

describe('ProfilePreferences component', () => {
  let wrapper;
  const defaultProvide = {
    languageChoices,
    firstDayOfWeekChoicesWithDefault,
    integrationViews: [],
    userFields,
    featureFlags: {},
  };

  function createComponent(options = {}) {
    const { props = {}, provide = {} } = options;
    return shallowMount(ProfilePreferences, {
      provide: {
        ...defaultProvide,
        ...provide,
      },
      propsData: props,
      stubs: {
        GlSprintf,
      },
    });
  }

  const findLocalizationAnchor = () => wrapper.find('#localization');

  const findUserLanguageOptionList = () =>
    wrapper.findAll('[data-testid="user-preferred-language-option"]');

  const findUserLanguageSelectedOption = () =>
    wrapper.find('[data-testid="user-preferred-language-option"]:checked');

  const findUserFirstDayOfWeekOptionList = () =>
    wrapper.findAll('[data-testid="user-first-day-of-week-option"]');

  const findUserFirstDayOfWeekSelectedOption = () =>
    wrapper.find('[data-testid="user-first-day-of-week-option"]:checked');

  const findUserTimeSettingsRule = () => wrapper.find('[data-testid="user-time-settings-rule"]');

  const findUserTimeSettingsHeading = () =>
    wrapper.find('[data-testid="user-time-settings-heading"]');

  const findUserTimeFormatOption = () => wrapper.find('[data-testid="user-time-format-option"]');

  const findUserTimeRelativeOption = () =>
    wrapper.find('[data-testid="user-time-relative-option"]');

  const findIntegrationsRule = () =>
    wrapper.find('[data-testid="profile-preferences-integrations-rule"]');

  const findIntegrationsHeading = () =>
    wrapper.find('[data-testid="profile-preferences-integrations-heading"]');

  const findIntegrationViewList = () => wrapper.findAll(IntegrationView);

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  describe('Localization Settings section', () => {
    beforeEach(() => {
      wrapper = createComponent();
    });

    it('has an id for anchoring', () => {
      expect(findLocalizationAnchor().exists()).toBe(true);
    });

    it('allows the user to change their language preferences', async () => {
      const newChoice = 1;
      const languageOptions = findUserLanguageOptionList();
      expect(findUserLanguageSelectedOption().element.value).toBe(languageChoices[0][1]);
      await languageOptions.at(newChoice).setSelected();
      expect(findUserLanguageSelectedOption().element.value).toBe(languageChoices[newChoice][1]);
    });

    it('allows the user to change their first day of the week preferences', async () => {
      const newChoice = 1;
      const languageOptions = findUserFirstDayOfWeekOptionList();
      expect(findUserFirstDayOfWeekSelectedOption().element.value).toBe(
        firstDayOfWeekChoicesWithDefault[0][1],
      );
      await languageOptions.at(newChoice).setSelected();
      expect(findUserFirstDayOfWeekSelectedOption().element.value).toBe(
        firstDayOfWeekChoicesWithDefault[newChoice][1],
      );
    });
  });

  describe('with `userTimeSettings` feature flag enabled', () => {
    beforeEach(() => {
      wrapper = createComponent({ provide: { featureFlags } });
    });

    it('should render user time settings', () => {
      expect(findUserTimeSettingsRule().exists()).toBe(true);
      expect(findUserTimeSettingsHeading().exists()).toBe(true);
      expect(findUserTimeFormatOption().exists()).toBe(true);
      expect(findUserTimeRelativeOption().exists()).toBe(true);
    });

    it('allows the user to toggle their time format preference', async () => {
      const userTimeFormatOption = findUserTimeFormatOption();
      expect(userTimeFormatOption.element.checked).toBe(false);
      await userTimeFormatOption.trigger('click');
      expect(userTimeFormatOption.element.checked).toBe(true);
    });

    it('allows the user to toggle their time display preference', async () => {
      const userTimeTimeRelativeOption = findUserTimeRelativeOption();
      expect(userTimeTimeRelativeOption.element.checked).toBe(false);
      await userTimeTimeRelativeOption.trigger('click');
      expect(userTimeTimeRelativeOption.element.checked).toBe(true);
    });
  });

  describe('with `userTimeSettings` feature flag disabled', () => {
    it('should not render user time settings', () => {
      wrapper = createComponent();
      expect(findUserTimeSettingsRule().exists()).toBe(false);
      expect(findUserTimeSettingsHeading().exists()).toBe(false);
      expect(findUserTimeFormatOption().exists()).toBe(false);
      expect(findUserTimeRelativeOption().exists()).toBe(false);
    });
  });

  describe('Integrations section', () => {
    it('should not render', () => {
      wrapper = createComponent();

      expect(findIntegrationsRule().exists()).toBe(false);
      expect(findIntegrationsHeading().exists()).toBe(false);
      expect(findIntegrationViewList()).toHaveLength(0);
    });

    it('should render', () => {
      wrapper = createComponent({ provide: { integrationViews } });

      expect(findIntegrationsRule().exists()).toBe(true);
      expect(findIntegrationsHeading().exists()).toBe(true);
      expect(findIntegrationViewList()).toHaveLength(integrationViews.length);
    });
  });

  it('should render ProfilePreferences properly', () => {
    wrapper = createComponent({
      provide: {
        integrationViews,
        featureFlags,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
