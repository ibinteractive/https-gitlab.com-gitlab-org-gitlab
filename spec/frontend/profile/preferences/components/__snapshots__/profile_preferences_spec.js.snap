// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilePreferences component should render ProfilePreferences properly 1`] = `
<div
  class="row gl-mt-3 js-preferences-form"
>
  <div
    class="col-sm-12"
  >
    <hr />
  </div>
   
  <div
    class="col-lg-4 profile-settings-sidebar"
  >
    <h4
      class="gl-mt-0"
    >
      
      Localization
    
    </h4>
     
    <p>
      
      Customize language and region related settings.
      
      <a
        href="/help/user/profile/preferences#localization"
        target="_blank"
      >
        
        Learn more 
      </a>
      .
    
    </p>
  </div>
   
  <div
    class="col-lg-8"
  >
    <div
      class="form-group"
    >
      <label
        class="label-bold"
        for="user_preferred_language"
      >
        
        Language
      
      </label>
       
      <div
        class="select-wrapper"
      >
        <select
          class="form-control select-control"
          id="user_preferred_language"
          name="user[preferred_language]"
          tabindex="-1"
          title="Language"
        >
          <option
            value="en"
          >
            
            English
          
          </option>
          <option
            value="da"
          >
            
            Danish
          
          </option>
          <option
            value="swe"
          >
            
            Swedish
          
          </option>
        </select>
         
        <i
          aria-hidden="true"
          class="fa fa-chevron-down"
        />
      </div>
       
      <div
        class="form-text text-muted"
      >
        
        This feature is experimental and translations are not complete yet
      
      </div>
    </div>
     
    <div
      class="form-group"
    >
      <label
        class="label-bold"
        for="user_first_day_of_week"
      >
        
        First day of the week
      
      </label>
       
      <div
        class="select-wrapper"
      >
        <select
          class="form-control select-control"
          id="user_first_day_of_week"
          name="user[first_day_of_week]"
          tabindex="-1"
          title="First day of the week"
        >
          <option
            value="1"
          >
            
            Monday
          
          </option>
          <option
            value="7"
          >
            
            Sunday
          
          </option>
        </select>
         
        <i
          aria-hidden="true"
          class="fa fa-chevron-down"
        />
      </div>
    </div>
  </div>
   
  <div
    class="col-sm-12"
    data-testid="user-time-settings-rule"
  >
    <hr />
  </div>
   
  <div
    class="col-lg-4 profile-settings-sidebar"
    data-testid="user-time-settings-heading"
  >
    <h4
      class="gl-mt-0"
    >
      
      Time preferences
    
    </h4>
     
    <p>
      
      These settings will update how dates and times are displayed for you.
    
    </p>
  </div>
   
  <div
    class="col-lg-8"
    data-testid="user-time-settings-option"
  >
    <h5>
      
      Time format
    
    </h5>
     
    <div
      class="form-group form-check"
    >
      <input
        name="user[time_format_in_24h]"
        type="hidden"
        value="0"
      />
       
      <input
        class="form-check-input"
        id="user_time_format_in_24h"
        name="user[time_format_in_24h]"
        type="checkbox"
        value="1"
      />
       
      <label
        class="form-check-label"
        for="user_time_format_in_24h"
      >
        
        Display time in 24-hour format
      
      </label>
    </div>
     
    <div
      class="form-group form-check"
    >
      <input
        name="user[time_display_relative]"
        type="hidden"
        value="0"
      />
       
      <input
        class="form-check-input"
        id="user_time_display_relative"
        name="user[time_display_relative]"
        type="checkbox"
        value="1"
      />
       
      <label
        class="form-check-label"
        for="user_time_display_relative"
      >
        
        For example: 30 mins ago.
      
      </label>
    </div>
  </div>
   
  <div
    class="col-sm-12"
  >
    <hr
      data-testid="profile-preferences-integrations-rule"
    />
  </div>
   
  <div
    class="col-lg-4 profile-settings-sidebar"
  >
    <h4
      class="gl-mt-0"
      data-testid="profile-preferences-integrations-heading"
    >
      
      Integrations
    
    </h4>
     
    <p>
      
      Customize integrations with third party services.
    
    </p>
  </div>
   
  <div
    class="col-lg-8"
  >
    <integration-view-stub
      config="[object Object]"
      helplink="http://foo.com/help"
      message="Click %{linkStart}Foo%{linkEnd}!"
      messageurl="http://foo.com"
    />
    <integration-view-stub
      config="[object Object]"
      helplink="http://bar.com/help"
      message="Click %{linkStart}Bar%{linkEnd}!"
      messageurl="http://bar.com"
    />
  </div>
</div>
`;
